src/bin/render-readmes.rs

```diff
@@ -2,87 +2,146 @@
-use cargo_registry::version::EncodableVersion;
 use cargo_registry::schema::*;
 use cargo_registry::render::markdown_to_html;
 
-const DEFAULT_PAGE_SIZE: i64 = 25;
+const DEFAULT_PAGE_SIZE: usize = 25;
+const USAGE: &'static str = "
+Usage: render-readmes [options]
+       render-readmes --help
+
+Options:
+    -h, --help         Show this message.
+    --page-size NUM    How many versions should be queried and processed at a time.
+    --older-than DATE  Only rerender readmes that are older than this date.
+    --crate NAME       Only rerender readmes for the specified crate.
+";
+
+#[derive(Deserialize)]
+struct Args {
+    flag_page_size: Option<usize>,
+    flag_older_than: Option<String>,
+    flag_crate: Option<String>,
+}
 
 fn main() {
+    let args: Args = Docopt::new(USAGE)
+        .and_then(|d| d.deserialize())
+        .unwrap_or_else(|e| e.exit());
     let config: Config = Default::default();
     let conn = cargo_registry::db::connect_now().unwrap();
-    let versions_count = versions::table
-        .select(versions::all_columns)
-        .count()
-        .get_result::<i64>(&conn)
-        .expect("error counting versions");
-    let page_size = match env::args().nth(1) {
-        None => DEFAULT_PAGE_SIZE,
-        Some(s) => s.parse::<i64>().unwrap_or(DEFAULT_PAGE_SIZE),
+
+    let start_time = Utc::now();
+
+    let older_than = if let Some(ref time) = args.flag_older_than {
+        Utc.datetime_from_str(&time, "%Y-%m-%d %H:%M:%S").expect(
+            "Could not parse --older-than argument as a time",
+        )
+    } else {
+        start_time
     };
-    let pages = if versions_count % page_size == 0 {
-        versions_count / page_size
+    let older_than = older_than.naive_utc();
+
+    println!("Start time:                   {}", start_time);
+    println!("Rendering readmes older than: {}", older_than);
+
+    let mut query = versions::table
+        .inner_join(readme_rendering::table)
+        .inner_join(crates::table)
+        .filter(readme_rendering::rendered_at.lt(older_than))
+        .select(versions::id)
+        .into_boxed();
+
+    if let Some(crate_name) = args.flag_crate {
+        println!("Rendering readmes for {}", crate_name);
+        query = query.filter(crates::name.eq(crate_name));
+    }
+
+    let version_ids = query.load::<(i32)>(&conn).expect(
+        "error loading version ids",
+    );
+
+    let total_versions = version_ids.len();
+    println!("Rendering {} versions", total_versions);
+
+    let page_size = args.flag_page_size.unwrap_or(DEFAULT_PAGE_SIZE);
+
+    let total_pages = total_versions / page_size;
+    let total_pages = if total_versions % page_size == 0 {
+        total_pages
     } else {
-        versions_count / page_size + 1
+        total_pages + 1
     };
-    for current_page in 0..pages {
-        let versions: Vec<EncodableVersion> = versions::table
+
+    let mut page_num = 0;
+
+    for version_ids_chunk in &version_ids.into_iter().chunks(page_size) {
```

Comment:
I didn't know about itertools, awesome crate!

Couldn't we use [enumerate](https://docs.rs/itertools/0.6.1/itertools/fn.enumerate.html) here to remove the `mut page_num` ?

Reply:
oooh yep, you're absolutely right!